### PR TITLE
Add per project config support

### PR DESCRIPTION
**Pull request by zaaack**

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+function joinEnvPath(...paths: (string | undefined)[]) {
+  return flatten1(paths)
+    .filter(Boolean)
+    .join(path.delimiter)
+}
+
+function readFileConf(dir: string, file: string, property?: string) {
+  try {
+    const conf = require(path.join(dir, 'package.json'))
```
Comment (clouds56) [inline src/main.ts]: here should be `path.join(dir, file)`

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      JSON.stringify(defaultConfig, null, 2),
+      'utf-8',
+    )
+  }
+
+  startServerProcess(projectPath: string) {
+    const serverPath = require.resolve('ocaml-language-server/bin/server')
+
+
+    const confFromPkg: Partial<Config> = readFileConf(projectPath, 'package.json', 'atom-reason') || {}
```
Comment (clouds56) [inline src/main.ts]: Is it really Partial<Config>? I think RecursivePartial is more accurate. see also https://stackoverflow.com/questions/41980195/recursive-partialt-in-typescript-2-1

Comment (zaaack) [reply src/main.ts]: Nice trick, but it's not working with deepmerge's tsd in this case.

**Review by alex35mil — COMMENTED** [review]
One question. There's `path` property in config that holds paths to different binaries. And AFAIU if those are not in the PATH, user must provide absolute path to each. In case of solo devs it's not an issue. But what about teams? I expect that `.atom/ide-reason.json` is commited and it holds paths to project's `node_modules` bin. Then the paths in generated config will work for the author of commit but other team members won't be able to use this config?

File: src/main.ts (hunk 2 above)
Comment (zaaack) [reply src/main.ts]: types/deepmerge has bugs about ts's default export, I'll take your advice!

**Comment by zaaack** [comment]
@alexfedoseev Are you mean we should provide relative binary paths in node_modules in generated config file instead of in env `PATH`s? I'm not sure most people are using this plugin in  BuckleScript projects or native projects, but env `PATH` config should be working in both kind. And they are same as atom's config panel's default values. Perhaps we can add some tips in documents?

**Comment by alex35mil** [comment]
@zaaack Well yeah, that's the question. Should we ask for paths relative to project's root or those should be absolute. If I will commit now my config, it's going to contain:

```
"bsb": "/Users/Alex/.../my-app/node_modules/bin/bsb"
```

and it won't work for anyone except me. If my understanding of how it works is correct :)

**Comment by zaaack** [comment]
@alexfedoseev I see, I think relative path should work, like:

```sh
"bsb": "./node_modules/bsb",
```

But not tested yet.

**Comment by alex35mil** [comment]
If it works, then all good!

**Comment by clouds56** [comment]
I could not even start with the new branch.
I've cloned the repo and checkout dev, then `apm link`.
After start atom, the config page of ide-reason could show but other things are all broken:
Outline shows "No outline available, Reason files do not currently support outlines".

In fact the extension seems to refuse to load until I open the config page of ide-reason.
(for the master branch, everything just works)

The file not loaded (until open the config page), no log output (even with debugLSP), the ocamlmerlin-server is not started. Nothing happens.

PS: I see yarn.lock in the folder and assuming use `yarn && yarn build` to install dependencies.
Am I missing some configuration?

**Comment by zaaack** [comment]
@alexfedoseev Just tested relative paths, it's working in ocaml-language-server.

@clouds56 It looks all good here, if you are OK, too then I think it's time for merging.

File: src/main.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import merge from 'deepmerge'
+import * as languageServer from 'ocaml-language-server'
+
+// let debug = (...args: any[]) => {
+//   console.log(...args)
+// }
+
+
+type DeepPartial<T> = {
+  [K in keyof T]?: DeepPartial<T>
```
Comment (clouds56) [inline src/main.ts]: [K in keyof T]?: DeepPartial<T[K]>

**Comment by clouds56** [comment]
No luck, it only works when I open ocaml file, but not reason file.

I see, I've installed language-reason but not language-reasonml.

File: package.json [hunk 4]
```diff
@@ -28,10 +32,18 @@
...
     "atom-package-deps": "^4.6.1",
+    "debug": "^3.1.0",
+    "deepmerge": "^2.0.1",
+    "fs-extra": "^5.0.0",
     "ocaml-language-server": "1.0.27"
   },
+  "activationCommands": {
+    "atom-workspace": "ide-reason:generateConfig"
+  },
+  "activationHooks": ["language-reasonml:grammar-used", "language-ocaml:grammar-used"],
```
Comment (clouds56) [inline package.json]: language-reason but not language-reasonml

File: src/main.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  confPerProject: DeepPartial<Config> = {}
+
+  getGrammarScopes () { return scopes }
+  getLanguageName () { return 'Reason' }
+  getServerName () { return 'ocamlmerlin' }
+  getConnectionType() { return 'ipc' as 'ipc' }
+  getRootConfigurationKey() { return 'ide-reason' }
+
+  mapConfigurationObject(config: Config) {
+    config = merge(merge({}, config), this.confPerProject)
```
Comment (clouds56) [inline src/main.ts:38]: this line fails compile? (I remember once it could even fixed the `DeepPartial<T[K]>` issue.)

**Comment by zaaack** [comment]
I think I just made some really stupid mistake... :sweat_smile:   Not too bad, it's right now.

File: src/main.ts (hunk 5 above)
Comment (zaaack) [reply src/main.ts:38]: It should be fixed now, I'm using a custom d.ts for `deepmerge`.

**Comment by clouds56** [comment]
Let's talk about `mapConfigurationObject`, I see it call the function is called after "ocamlmerlin-server" has been started, and the new "ocamlmerlin" path in the config does not take effect, still use the one in system path.
Should/could we pass the config at server spawn (in `startServerProcess`)?

File: package.json [hunk 6]
```diff
@@ -28,10 +32,16 @@
...
+    "atom-languageclient": "^0.8.3",
     "atom-package-deps": "^4.6.1",
+    "deepmerge": "^2.0.1",
+    "fs-extra": "^5.0.0",
     "ocaml-language-server": "1.0.27"
   },
+  "activationCommands": {
+    "atom-workspace": "ide-reason:generate-config"
+  },
+  "activationHooks": ["language-reasonml:grammar-used", "language-ocaml:grammar-used"],
```
Comment (alex35mil) [inline package.json]: `"language-reasonml:grammar-used" -> "language-reason:grammar-used"`

Without `ml`.

**Review by alex35mil — CHANGES_REQUESTED** [review]
I pulled the latest, there's one typo. After this fix reason tooling worked for me.

File: package.json (hunk 6 above)
Comment (zaaack) [reply package.json]: Just finded in another pr...I'll fix it several hours later.

**Comment by alex35mil** [comment]
> Let's talk about mapConfigurationObject, I see it call the function is called after "ocamlmerlin-server" has been started, and the new "ocamlmerlin" path in the config does not take effect, still use the one in system path.
> Should/could we pass the config at server spawn (in startServerProcess)?

AFAIR `atom-languageclient` should send message to `ocaml-language-server` on config change and the latter should use the latest.

**Comment by zaaack** [comment]
> @clouds56 Should/could we pass the config at server spawn (in `startServerProcess`)?

@clouds56 @alexfedoseev Just added `didChangeConfiguration` events and restartAllServer command.

**Comment by alex35mil** [comment]
I just tested local config and it works! GJ 👍 

@zaaack @clouds56 BTW do you guys see `ocamlmerlin error: Fatal error: exception Sys_error("Broken pipe")`? https://cl.ly/pyXw

It pops up often on save. I saw the same error on VSCode but it was not that often.

**Comment by zaaack** [comment]
@alexfedoseev I can remember something like this happens very often when I was still using atom heavily (now more is vsc + ts + [fable](https://github.com/fable-compiler/Fable) ), that's why I asked the atom-languageclient's maintainer to add the restart servers feature(https://github.com/atom/atom-languageclient/issues/71).  I guess it's because the merlin process loss response when you edit too much files.

File: src/main.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    this.subscriptions.add(
+      atom.commands.add('atom-workspace', {
+        [`${pkg.name}:generate-config`]: () => this.generateConfig().catch(console.error)
+      }),
+      atom.commands.add('atom-workspace', {
+        [`${pkg.name}:restart-all-servers`]: () => this.restartAllServers().catch(console.error)
+      }),
+    )
+
+    const handleConfFileChange = throttle((events: FilesystemChangeEvent) => {
```
Comment (alex35mil) [inline src/main.ts]: BTW why do you throttle here? AFAIU `atom.project.onDidChangeFiles` notifies us on change in fs (when user saves file) so `handleConfFileChange` is one time operation we can fire right away?

Comment (zaaack) [reply src/main.ts]: You are right, it's not necessary...

**Review by alex35mil — APPROVED** [review]
LGTM 👍

**Review by clouds56 — APPROVED** [review]
LGTM

**Comment by zaaack** [comment]
@alexfedoseev @clouds56 Thanks a lot! `ide-reason@v0.1.12` just published! 🎉 🎉

**Pull request merged**